backend/copr_backend/pulp.py

```diff
@@ -10,6 +10,148 @@
+        """
+        Drop our entry from Redis DB (if any), and return True if the task is
+        already processed.  When 'delete_if_not=True, we delete the self.key
+        from Redis even if the task is not yet processed (meaning that caller
+        plans to finish the task right away).
+        """
+        if self.noop:
+            return False
+
+        status = self.redis.hget(self.key, "status") == "success"
+        self.log.debug("Has already a status? %s", status)
+
+        try:
+            if not status:
+                # not yet processed
+                return False
+        finally:
+            # This is atomic operation, other processes may not re-start doing this
+            # task again.  https://github.com/redis/redis/issues/9531
+            #
+            # The 'delete_if_not = True' is not a TOCTOU case.  Our caller holds
+            # the lock, hence no other machine may take our task (would require
+            # lock) and re-process it.
+            if status or delete_if_not:
+                self.redis.delete(self.key)
+
+        return status
+
+    def options(self):
+        """
+        Get the options from other _compatible_ (see below) Redis tasks, and
+        plan the list of tasks in self.notify_keys[] that we will notify in
+        commit().
+        """
+        rpms_to_add = set(self.rpms_to_add)
+        rpms_to_remove = set(self.rpms_to_remove)
+
+        if self.noop:
+            return {"add_content_units": rpms_to_add, "remove_content_units": rpms_to_remove}
+
+        for key in self.redis.keys(self.key_pattern):
+            assert key != self.key
+
+            task_dict = self.redis.hgetall(key)
+            if not task_dict:
+                # TOCTOU: The key might no longer exist in DB, even though
+                # _our process_ holds the lock!  See the
+                # 'self.redis.delete(self.key)' above in check_processed(); the
+                # _original process_ removes key without locking, if _third_
+                # process commit()ted the status=succeeded.
+                # Prior fixing #3770/#3777 we tried to process these deleted
+                # tasks and ended up with KeyError on task_dict["task"] below.
+                self.log.info("Key %s already processed (by other process), "
+                              "removed by the original process (without lock) "
+                              "since we listed it.", key)
+                continue
+
+            if task_dict.get("status") is not None:
+                # skip processed tasks
+                self.log.info("Key %s already processed, skip", key)
+                continue
+
+            task_opts = json.loads(task_dict["task"])
+
+            # we can process this task!
+            self.notify_keys.append(key)
+
+            # append "add" tasks, if that makes sense
+            rpms_to_add.update(task_opts["add_content_units"])
+            rpms_to_remove.update(task_opts["remove_content_units"])
```

Comment:
Note three's a chance that the same prn may appear in both sets, so perhaps we shouldn't request the same thing to be added and removed in one API call?

Also, you can use set operators like `rpms_to_add |= set(task_opts["add_content_units"])`, and

```
common = rpms_to_add & rpms_to_del
rpms_to_add -= common
rpms_to_del -= common
```
```